## Arithmetic operators

In [11]:
x, y = 5, 3

In [12]:
s = x + y
d = x - y
p = x * y
q = x / y
qf = x // y
r = x % y
e = x ** y

In [25]:
print('x =', x)
print('y =', y)
print('sum        =', s)
print('difference =', d)
print('product    =', p)
print('divide     =', q)
print('quotient   =', qf)
print('remainder  =', r)
print('power      =', e)

x = 5
y = 3
sum        = 8
difference = 2
product    = 15
divide     = 1.6666666666666667
quotient   = 1
remainder  = 2
power      = 125


<hr>

## Relational operators

In [26]:
print('x =', x)
print('y =', y)

x = 5
y = 3


In [27]:
x > y

True

In [28]:
x < y

False

In [20]:
x != y

True

In [29]:
10 == 10

True

In [30]:
2 != 1

True

In [32]:
5 >= 5

True

## Chaining up of comparison operators
If there are multiple operations connected with a `logical and`, they can be chained together into one expression.

In [61]:
x = 56

### Check if `x` lies within 0 and 100

In [62]:
if x > 0 and x < 100:
    print('yes')
else:
    print('no')

yes


In [64]:
if 0 < x < 100:
    print('yes')
else:
    print('no')

yes


***

## Shorthand assignment operators

In [33]:
x = 10
x += 3  # this translates to x = x + 3
print(x)

13


In [35]:
x = 20
x *= 3
print(x)
x -= 50
print(x)

60
10


***

## Logical operators

In [36]:
x, y = 2, 3

In [45]:
x != y and x < y and x+y > 0

True

In [44]:
x != y and x > y

False

In [42]:
x == y or x*10 == y or y-x == 1

True

In [41]:
not x != y

False

***

### Quick question
What is the value of <font color="red">2 * 5 ** 3</font> ?
1. 1000
2. 250
3. I've never seen ** before! It's a syntax error

In [49]:
2 * 5 ** 3

250

## The importance of operator precedence

<table>
    <tr>
        <th><font size=5>Symbol</font></th>
        <th><font size=5>Operator</font></th>
    </tr>
    <tr>
        <td><font size=4>( )</font></td>
        <td><font size=4>Parentheses</font></td>
    </tr>
    <tr>
        <td><font size=4>**</font></td>
        <td><font size=4>Exponent</font></td>
    </tr>
    <tr>
        <td><font size=4>+x, -x, ~x</font></td>
        <td><font size=4>Unary plus, Unary minus, Bitwise complement</font></td>
    </tr>
    <tr>
        <td><font size=4>*, /, //, %</font></td>
        <td><font size=4>Multiplication, Division, Floor division, Modulus</font></td>
    </tr>
    <tr>
        <td><font size=4>+, -</font></td>
        <td><font size=4>Addition, Subtraction</font></td>
    </tr>
    <tr>
        <td><font size=4>&</font></td>
        <td><font size=4>Bitwise AND</font></td>
    </tr>
    <tr>
        <td><font size=4>^</font></td>
        <td><font size=4>Bitwise XOR</font></td>
    </tr>
    <tr>
        <td><font size=4>|</font></td>
        <td><font size=4>Bitwise OR</font></td>
    </tr>
    <tr>
        <td><font size=4>==, !=, >, >=, <, <=, is, is not, in, not in</font></td>
        <td><font size=4>Comparisions, Identity, Membership operators</font></td>
    </tr>
    <tr>
        <td><font size=4>not</font></td>
        <td><font size=4>Logical NOT</font></td>
    </tr>
    <tr>
        <td><font size=4>and</font></td>
        <td><font size=4>Logical AND</font></td>
    </tr>
    <tr>
        <td><font size=4>or</font></td>
        <td><font size=4>Logical OR</font></td>
    </tr>
</table>

***

### Quick question
What is the value of <font color="red">2 ** 3 ** 4</font> ?
1. 2417851639229258349412352
2. 4096
3. This should be illegal

In [55]:
2 ** 3 ** 4

2417851639229258349412352

In [56]:
(2 ** 3) ** 4

4096

## The importance of operator associativity

We can see that there are multiple operators in one group in the precedence table.  

Associativity helps in determining the order of operations when two or more operators have the same precedence.  

The associativity of the ** operator is from right to left.

### Quick question
What is the value of <font color="red">5 * 2 // 3</font> ?
1. 0
2. 3
3. I'd rather not answer

In [3]:
5 * 2 // 3

3

In [4]:
5 * (2 // 3)

0

The other operators have left to right associativity

***

### Problems with floating point math

#### What is 0.1 + 0.2 ?
#### We all know that 0.1 + 0.2 is equal to 0.3.

In [8]:
.1 + .2 == .3

False

In [2]:
.1 + .2

0.30000000000000004

### This website gives a brief description of why this happens
[0.30000000000000004.com](0.30000000000000004.com)

### How can we fix this problem?

The `math` module in python provides a function called `isclose`

In [5]:
from math import isclose

help(isclose)

Help on built-in function isclose in module math:

isclose(a, b, *, rel_tol=1e-09, abs_tol=0.0)
    Determine whether two floating point numbers are close in value.
    
      rel_tol
        maximum difference for being considered "close", relative to the
        magnitude of the input values
      abs_tol
        maximum difference for being considered "close", regardless of the
        magnitude of the input values
    
    Return True if a is close in value to b, and False otherwise.
    
    For the values to be considered close, the difference between them
    must be smaller than at least one of the tolerances.
    
    -inf, inf and NaN behave similarly to the IEEE 754 Standard.  That
    is, NaN is not close to anything, even itself.  inf and -inf are
    only close to themselves.



In [6]:
isclose(.1 + .2, .3)

True

***

![questions.jpg](img/questions.jpg)